README.md
### Lil'Blocks, a block programming language

Lil'Blocks is a derviative of [BlocklyDuino](https://github.com/gasolin/BlocklyDuino).

[BlocklyDuino](https://github.com/gasolin/BlocklyDuino) is a web-based visual programming editor for [Arduino](http://www.arduino.cc/).

[BlocklyDuino](https://github.com/gasolin/BlocklyDuino) is based on [Blockly](http://code.google.com/p/blockly/), the web-based, graphical programming editor.

### Authors and Contributors

Fred Lin (@gasolin), author of [BlocklyDuino](https://github.com/gasolin/BlocklyDuino)(https://github.com/gasolin/BlocklyDuino)

Neil Fraser, Q.Neutron, authors of Blockly http://code.google.com/p/blockly/

The project is also inspired by ardublock https://github.com/taweili/ardublock and modkit http://www.modk.it/

Chris Hakim (@chrishakim), Lil'Blocks adaptation of BlocklyDuino, in progress

### License

Copyright (C) 2014 Chris Hakim

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0


CHANGELOG.txt
Lil’Blocks
==========

2014-10-20
Procedure declaration problem fixed: function parameters were declared without type, and
global variables of int type were created. If procedure returned a value, procedure return type
was declared as Dynamic, which Arduino did not want to compile.
There is now a superficial fix of always declaring function parameters as int type and return
type as either void or int. This works, but redundant global variables are also created. This
may be confusing for beginners.

A block has been renamed from “rotate” to “turn” to be clearer to young programmers. The
Arduino rotate method keeps the same name.

2014-10-18
Initial robot blocks created

BlocklyDuino
=======================

1.0.1 2013/
------------------
code name:

Feature: 

Change:
* separate lcd include statement
* update blockly till r546
* rename package demos to apps

Fix:

Credit: gasolin

1.0 2012/12/11
------------------
code name: Mindstorms

Feature: 

* add procedure support with return and logic null blocks, contribute back to blockly
* add blocks:
  Grove Ultrasonic Ranger
  Grove Motor Shield
  Grove Thumb JoyStick
  Grove Chained RGB LED

Change:
* clean blocks internal
* update blockly till r518, pretty print xml
* change procedure return to ifreturn block via blockly

Fix:
* names with object
* variable dropdown with object

Credit: gasolin, neil fraser(blockly)

0.9 2012/11/11
------------------
code name: TECHNIC

Refactor blocks, rearrange import blocks and blocks with type checking

Feature: 

* add text blocks
* add blocks:
  Grove serial lcd effect
  Grove Sound Sensor
  Grove PIR Motion sensor
  Grove Line finder

Change:

* separate Grove category into Grove(digital), Grove Analog, and Grove LCD
* make LCD allow to input text block
* update blockly till r504, support checkbox
* remove jquery dependency by implement ajax via XMLHttpRequest
* rearrange import blocks

Fix:
* appcache support

Credit: gasolin

0.8 2012/10/31
------------------
code name: DUPLO

make editor more friendly

Feature: 

* generate arduino code in textarea, easier to copy source
* auto focus to arduino code area, can use ctrl+a to select correct source
* load source via url param in same site with such "url=xxx" syntax
* include 2 examples
* upload potentiometer control servo video http://www.youtube.com/watch?v=u2Fp0KpWxik
* application cache block images and jquery
* add Grove LCD print, Grove LCD power block

Change:

* call profile params via 'profile.default.digital' instead of 'profile.arduino.digital'
* rename Potentiometer block to Rotary Angle block
* update blockly till r496, move core js into core folder
* move arduino language generator to top level generator/ folder
* reasonable default) change default delay time from 0 to 1000

Credit: gasolin


0.7 2012/10/18
------------------
code name: SPACE

refine bootstraping process

Feature: 

* promotion video http://www.youtube.com/watch?v=_swiyXcUvNY
* bootstraping sample blocks and document to easily custom blocks.

Change:

* update blockly till 10/18
* from blockly) Move include section always at top of the source file
* move delay block to Control blocks Panel
* move map block to Math blocks panel

fix:

* fix servo move block cascade input issue 

Credit: gasolin


0.6 2012/10/14
-----------------
code name: CASTLE

make BlocklyDuino public to github

Feature:

* code on github
* add grove_temporature_sensor block
* add Serial print block

Change:

* map function can specify the max range value

Refactor:

Fix:

* fix several functions to able to assigned to variables

Credit: gasolin


0.5 2012/10/11
-----------------
code name: CREATOR

support int variables and new blockly syntax

Feature:

* support variable declare statement
* support type declaration for strong type language
  * (currently only support int type)
* can restore blocks
* add for statement

Refactor:

* add board profiles to support different board with ports(current arduino standard-compatible board only)
* support new blockly syntax
* change image size to 64*64

Credit: gasolin


0.4 2012/10/6
-----------------
code name: CITY

First version with minimum function sets.

Feature:

* Compact Digital/Analog Read/Write syntax
* Auto setup pinMode
* delay and map function
* 2 Standard blocks: Build-in LED, Servo 
* 6 Grove blocks: LED, Potentiometer, Tilt switch, Piezo Buzzer, Relay

Credit: gasolin


Code name convention
------------------------
BlocklyDuino code names follows lego themes http://en.wikipedia.org/wiki/Lego

License
-----------------
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0


blockly/generators/arduino/base.js
/**
 * Visual Blocks Language
 *
 * Copyright 2012 Fred Lin.
 * https://github.com/gasolin/BlocklyDuino
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @fileoverview Helper functions for generating Arduino blocks.
 * @author gasolin@gmail.com (Fred Lin)
 */
'use strict';

//To support syntax defined in http://arduino.cc/en/Reference/HomePage

//define blocks
if (!Blockly.Language) Blockly.Language = {};

Blockly.Language.base_delay = {
  category: 'Control',
  helpUrl: 'http://arduino.cc/en/Reference/delay',
  init: function() {
    this.setColour(120);
    this.appendValueInput("DELAY_TIME", Number)
        .appendTitle("Delay")
        .setCheck(Number);
    this.setInputsInline(true);
    this.setPreviousStatement(true, null);
    this.setNextStatement(true, null);
    this.setTooltip('Delay specific time');
  }
};

Blockly.Language.base_map = {
  category: 'Math',
  helpUrl: 'http://arduino.cc/en/Reference/map',
  init: function() {
    this.setColour(230);
    this.appendValueInput("NUM", Number)
        .appendTitle("Map ")
        .setCheck(Number);
    this.appendValueInput("DMAX", Number)
        .appendTitle("value to [0-")
        .setCheck(Number);
    this.appendDummyInput("")
	      .appendTitle("]");
    this.setInputsInline(true);
    this.setOutput(true);
    this.setTooltip('Re-maps a number from [0-1024] to another.');
  }
};

Blockly.Language.inout_buildin_led = {
   category: 'In/Out',
   helpUrl: 'http://arduino.cc/en/Reference/DigitalWrite',
   init: function() {
     this.setColour(190);
     this.appendDummyInput("")
	       .appendTitle("Build-in LED Stat")
	       .appendTitle(new Blockly.FieldDropdown([["HIGH", "HIGH"], ["LOW", "LOW"]]), "STAT");
     this.setPreviousStatement(true, null);
     this.setNextStatement(true, null);
     this.setTooltip('light or off the build-in LED');
   }
};

Blockly.Language.inout_digital_write = {
  category: 'In/Out',
  helpUrl: 'http://arduino.cc/en/Reference/DigitalWrite',
  init: function() {
    this.setColour(230);
    this.appendDummyInput("")
	      .appendTitle("DigitalWrite PIN#")
	      .appendTitle(new Blockly.FieldDropdown(profile.default.digital), "PIN")
      	.appendTitle("Stat")
      	.appendTitle(new Blockly.FieldDropdown([["HIGH", "HIGH"], ["LOW", "LOW"]]), "STAT");
    this.setPreviousStatement(true, null);
    this.setNextStatement(true, null);
    this.setTooltip('Write digital value to a specific Port');
  }
};

Blockly.Language.inout_digital_read = {
  category: 'In/Out',
  helpUrl: 'http://arduino.cc/en/Reference/DigitalRead',
  init: function() {
    this.setColour(230);
    this.appendDummyInput("")
	      .appendTitle("DigitalRead PIN#")
	      .appendTitle(new Blockly.FieldDropdown(profile.default.digital), "PIN");
    this.setOutput(true, Boolean);
    this.setTooltip('');
  }
};

Blockly.Language.inout_analog_write = {
  category: 'In/Out',
  helpUrl: 'http://arduino.cc/en/Reference/AnalogWrite',
  init: function() {
    this.setColour(230);
    this.appendDummyInput("")
        .appendTitle("AnalogWrite PIN#")
        .appendTitle(new Blockly.FieldDropdown(profile.default.pwmOutput), "PIN");
    this.appendValueInput("NUM", Number)
        .appendTitle("value")
        .setCheck(Number);
    this.setInputsInline(true);
    this.setPreviousStatement(true, null);
    this.setNextStatement(true, null);
    this.setTooltip('Write analog value between 0 and 255 to a specific Port');
  }
};

Blockly.Language.inout_analog_read = {
  category: 'In/Out',
  helpUrl: 'http://arduino.cc/en/Reference/AnalogRead',
  init: function() {
    this.setColour(230);
    this.appendDummyInput("")
        .appendTitle("AnalogRead PIN#")
        .appendTitle(new Blockly.FieldDropdown(profile.default.analog), "PIN");
    this.setOutput(true, Number);
    this.setTooltip('Return value between 0 and 1024');
  }
};

Blockly.Language.inout_highlow = {
  category: 'In/Out',
  helpUrl: 'http://arduino.cc/en/Reference/Constants',
  init: function() {
    this.setColour(230);
    this.appendDummyInput("")
        .appendTitle(new Blockly.FieldDropdown([["HIGH", "HIGH"], ["LOW", "LOW"]]), 'BOOL')
    this.setOutput(true, Boolean);
    this.setTooltip(Blockly.LANG_LOGIC_BOOLEAN_TOOLTIP_1);
  }
};

//servo block
//http://www.seeedstudio.com/depot/emax-9g-es08a-high-sensitive-mini-servo-p-760.html?cPath=170_171
Blockly.Language.servo_move = {
  category: 'Servo',
  helpUrl: 'http://www.arduino.cc/playground/ComponentLib/servo',
  init: function() {
    this.setColour(190);
    this.appendDummyInput("")
        .appendTitle("Servo")
        .appendTitle(new Blockly.FieldImage("http://www.seeedstudio.com/depot/images/product/a991.jpg", 64, 64))
        .appendTitle("PIN#")
        .appendTitle(new Blockly.FieldDropdown(profile.default.digital), "PIN")
    this.appendValueInput("DEGREE", Number)
        .setCheck(Number)
        .setAlign(Blockly.ALIGN_RIGHT)
        .appendTitle("Degree (0~180)");
    this.appendValueInput("DELAY_TIME", Number)
        .setCheck(Number)
        .setAlign(Blockly.ALIGN_RIGHT)
        .appendTitle("Delay");
    this.setPreviousStatement(true, null);
    this.setNextStatement(true, null);
    this.setTooltip('move between 0~180 degree');
  }
};

Blockly.Language.servo_read_degrees = {
  category: 'Servo',
  helpUrl: 'http://www.arduino.cc/playground/ComponentLib/servo',
  init: function() {
    this.setColour(190);
    this.appendDummyInput("")
        .appendTitle("Servo")
        .appendTitle(new Blockly.FieldImage("http://www.seeedstudio.com/depot/images/product/a991.jpg", 64, 64))
        .appendTitle("PIN#")
        .appendTitle(new Blockly.FieldDropdown(profile.default.digital), "PIN");
    this.appendDummyInput("")
        .setAlign(Blockly.ALIGN_RIGHT)
        .appendTitle("Read Degrees")
    this.setOutput(true, Number);
    this.setTooltip('return that degree with the last servo move.');
  }
};

Blockly.Language.serial_print = {
  category: 'In/Out',
  helpUrl: 'http://www.arduino.cc/en/Serial/Print',
  init: function() {
    this.setColour(230);
    this.appendValueInput("CONTENT", String)
        .appendTitle("Serial Print");
    this.setPreviousStatement(true, null);
    this.setNextStatement(true, null);
    this.setTooltip('Prints data to the console/serial port as human-readable ASCII text.');
  }
};

// define generators
Blockly.Arduino = Blockly.Generator.get('Arduino');

Blockly.Arduino.base_delay = function() {
  var delay_time = Blockly.Arduino.valueToCode(this, 'DELAY_TIME', Blockly.Arduino.ORDER_ATOMIC) || '1000'
  delay_time = delay_time.replace('(','').replace(')','');
  var code = 'delay(' + delay_time + ');\n';
  return code;
};

Blockly.Arduino.base_map = function() {
  var value_num = Blockly.Arduino.valueToCode(this, 'NUM', Blockly.Arduino.ORDER_NONE);
  var value_dmax = Blockly.Arduino.valueToCode(this, 'DMAX', Blockly.Arduino.ORDER_ATOMIC);
  //value_num = value_num.replace('(','').replace(')','');
  value_dmax = value_dmax.replace('(','').replace(')','');
  var code = 'map('+value_num+', 0, 1024, 0, '+value_dmax+')';
  return [code, Blockly.Arduino.ORDER_NONE];
};

Blockly.Arduino.inout_buildin_led = function() {
  var dropdown_stat = this.getTitleValue('STAT');
  Blockly.Arduino.setups_['setup_output_13'] = 'pinMode(13, OUTPUT);';
  var code = 'digitalWrite(13,'+dropdown_stat+');\n'
  return code;
};

Blockly.Arduino.inout_digital_write = function() {
  var dropdown_pin = this.getTitleValue('PIN');
  var dropdown_stat = this.getTitleValue('STAT');
  Blockly.Arduino.setups_['setup_output_'+dropdown_pin] = 'pinMode('+dropdown_pin+', OUTPUT);';
  var code = 'digitalWrite('+dropdown_pin+','+dropdown_stat+');\n'
  return code;
};

Blockly.Arduino.inout_digital_read = function() {
  var dropdown_pin = this.getTitleValue('PIN');
  Blockly.Arduino.setups_['setup_input_'+dropdown_pin] = 'pinMode('+dropdown_pin+', INPUT);';
  var code = 'digitalRead('+dropdown_pin+')';
  return [code, Blockly.Arduino.ORDER_ATOMIC];
};

Blockly.Arduino.inout_analog_write = function() {
  var dropdown_pin = this.getTitleValue('PIN');
  //var dropdown_stat = this.getTitleValue('STAT');
  var value_num = Blockly.Arduino.valueToCode(this, 'NUM', Blockly.Arduino.ORDER_ATOMIC);
  value_num = value_num.replace('(','').replace(')','');
  //Blockly.Arduino.setups_['setup_output'+dropdown_pin] = 'pinMode('+dropdown_pin+', OUTPUT);';
  var code = 'analogWrite('+dropdown_pin+','+value_num+');\n';
  return code;
};

Blockly.Arduino.inout_analog_read = function() {
  var dropdown_pin = this.getTitleValue('PIN');
  //Blockly.Arduino.setups_['setup_input_'+dropdown_pin] = 'pinMode('+dropdown_pin+', INPUT);';
  var code = 'analogRead('+dropdown_pin+')';
  return [code, Blockly.Arduino.ORDER_ATOMIC];
};

Blockly.Arduino.inout_highlow = function() {
  // Boolean values HIGH and LOW.
  var code = (this.getTitleValue('BOOL') == 'HIGH') ? 'HIGH' : 'LOW';
  return [code, Blockly.Arduino.ORDER_ATOMIC];
};

/*
//servo
#include <Servo.h>

Servo servo_11;

void setup() { 
  servo_11.attach(11);
}

void loop() { 
servo_11.write(0);
delay(2000);

servo_11.write(150); //0~180
delay(2000);
}
*/
Blockly.Arduino.servo_move = function() {
  var dropdown_pin = this.getTitleValue('PIN');
  var value_degree = Blockly.Arduino.valueToCode(this, 'DEGREE', Blockly.Arduino.ORDER_ATOMIC);
  //value_degree = value_degree.replace('(','').replace(')','')
  var delay_time = Blockly.Arduino.valueToCode(this, 'DELAY_TIME', Blockly.Arduino.ORDER_ATOMIC) || '1000'
  //delay_time = delay_time.replace('(','').replace(')','');
  
  Blockly.Arduino.definitions_['define_servo'] = '#include <Servo.h>\n';
  Blockly.Arduino.definitions_['var_servo'+dropdown_pin] = 'Servo servo_'+dropdown_pin+';\n';
  Blockly.Arduino.setups_['setup_servo_'+dropdown_pin] = 'servo_'+dropdown_pin+'.attach('+dropdown_pin+');\n';
  
  var code = 'servo_'+dropdown_pin+'.write('+value_degree+');\n'+'delay(' + delay_time + ');\n';
  return code;
};

Blockly.Arduino.servo_read_degrees = function() {
  var dropdown_pin = this.getTitleValue('PIN');
  
  Blockly.Arduino.definitions_['define_servo'] = '#include &lt;Servo.h&gt;\n';
  Blockly.Arduino.definitions_['var_servo'+dropdown_pin] = 'Servo servo_'+dropdown_pin+';\n';
  Blockly.Arduino.setups_['setup_servo_'+dropdown_pin] = 'servo_'+dropdown_pin+'.attach('+dropdown_pin+');\n';
  
  var code = 'servo_'+dropdown_pin+'.read()';
  return code;
};

Blockly.Arduino.serial_print = function() {
  var content = Blockly.Arduino.valueToCode(this, 'CONTENT', Blockly.Arduino.ORDER_ATOMIC) || '0'
  content = content.replace('(','').replace(')','');
  
  // Setup already happening in Bot.begin()
  // Blockly.Arduino.setups_['setup_serial_'+profile.default.serial] = 'Serial.begin('+profile.default.serial+');\n';
  
  var code = 'Serial.print('+content+');\n';
  return code;
};
